### PR TITLE
initial support for pruning stale entries from the metricIndex

### PR DESCRIPTION
**Pull request by woodsaj**
Periodically prune entries from the meticIndex if they are stale

### CODE REVIEW

File: idx/memory/memory.go [hunk 1]
```diff
@@ -518,6 +518,57 @@ func (m *MemoryIdx) delete(orgId int, n *Node) ([]string, error) {
...
+// delete metricDefs from the index if they have not been seen since "oldest"
+func (m *MemoryIdx) Prune(orgId int, oldest time.Time) ([]schema.MetricDefinition, error) {
+	oldestUnix := oldest.Unix()
+	pruned := make([]schema.MetricDefinition, 0)
+	pre := time.Now()
+	m.RLock()
+	defer m.RUnlock()
+	orgs := []int{orgId}
+	if orgId == -1 {
+		log.Info("pruning stale metricDefs across all orgs")
```
Comment (Dieterbe) [inline idx/memory/memory.go]: logging calls should have a prefix

File: idx/memory/memory.go [hunk 2]
```diff
@@ -518,6 +518,57 @@ func (m *MemoryIdx) delete(orgId int, n *Node) ([]string, error) {
...
+				//we need to delete this node.
+				_, err := m.delete(org, n)
+				if err != nil {
+					return pruned, err
+				}
+			}
+		}
+	}
+	if orgId == -1 {
+		log.Info("pruning stale metricDefs from memory for all orgs took %s", time.Since(pre).String())
```
Comment (Dieterbe) [inline idx/memory/memory.go]: log prefix

File: idx/memory/memory.go [hunk 3]
```diff
@@ -518,6 +518,57 @@ func (m *MemoryIdx) delete(orgId int, n *Node) ([]string, error) {
...
+			continue
+		}
+		for _, n := range tree.Items {
+			if !n.Leaf {
+				continue
+			}
+			for i := len(n.Children) - 1; i >= 0; i-- {
+				id := n.Children[i]
+				if m.DefById[id].LastUpdate < oldestUnix {
+					log.Info("memoryIdx: metricDef %s is stale. pruning it.", id)
```
Comment (Dieterbe) [inline idx/memory/memory.go]: shouldn't this be a debug level call?

File: idx/memory/memory.go [hunk 4]
```diff
@@ -518,6 +518,57 @@ func (m *MemoryIdx) delete(orgId int, n *Node) ([]string, error) {
...
+	for _, org := range orgs {
+		tree, ok := m.Tree[org]
+		if !ok {
+			continue
+		}
+		for _, n := range tree.Items {
+			if !n.Leaf {
+				continue
+			}
+			for i := len(n.Children) - 1; i >= 0; i-- {
```
Comment (Dieterbe) [inline idx/memory/memory.go]: just curious, why iterate in reverse order?

Comment (woodsaj) [inline idx/memory/memory.go]: If you delete from a slice while iterating over it, you are not going to get the result you expected as the positions of the items in the slice change when you delete an item.

https://play.golang.org/p/m5Pleaw2Os

**Pull request merged**